### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by Koliras**
[DEMO LINK](https://Koliras.github.io/react_dynamic-list-of-todos/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,14 +1,77 @@
...
+    case 'completed': {
+      filteredTodos = todosToFilter.filter(todo => todo.completed);
+      break;
+    }
+
+    default:
+      filteredTodos = todosToFilter;
+  }
+
+  filteredTodos = filteredTodos.filter(todo => todo.title.toLowerCase().includes(query.toLowerCase()));
```
Comment (rostyslav48) [inline src/App.tsx:36]: it is better to trim query `(query.toLowerCase().trim())` in this way we will remove extra spaces and improve our search system

File: src/App.tsx [hunk 2]
```diff
@@ -1,14 +1,77 @@
...
+
+  return filteredTodos;
+}
 
 export const App: React.FC = () => {
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+  const [loading, setLoading] = useState(false);
```
Comment (rostyslav48) [inline src/App.tsx:43]: You are using a boolean value for this var, so consider naming it something like `isLoading`

File: src/components/TodoFilter/TodoFilter.tsx [hunk 3]
```diff
@@ -1,8 +1,22 @@
-export const TodoFilter = () => (
+import React from 'react';
+
+type Props = {
+  onFilterKeyChange: (key: string) => void,
+  onQueryChange: (event: React.ChangeEvent<HTMLInputElement>) => void,
+  onDelete: () => void
+  query: string,
```
Comment (rostyslav48) [inline src/components/TodoFilter/TodoFilter.tsx:7]: It is really cool that you added some optimization features. Although, you pass some props which are not wrapped in `useMemo()`. The question is whether this optimization features needed.

File: src/components/TodoModal/TodoModal.tsx [hunk 4]
```diff
@@ -15,35 +35,40 @@ export const TodoModal: React.FC = () => {
...
           <div className="modal-card-body">
             <p className="block" data-cy="modal-title">
-              quis ut nam facilis et officia qui
+              {selectedTodo.title}
             </p>
 
             <p className="block" data-cy="modal-user">
               {/* <strong className="has-text-success">Done</strong> */}
-              <strong className="has-text-danger">Planned</strong>
+              <strong className="has-text-danger">
```
Comment (rostyslav48) [inline src/components/TodoModal/TodoModal.tsx:57]: You haven't changed color for the done tasks
Знімок екрана 2023-09-14 о 22 14 19

**Review by rostyslav48 — CHANGES_REQUESTED** [review]
Well Done👍
However there is a tiny problem with your app please check the last comment and fix it. All other comments are just a suggestions, so you may take them into consideration)

File: src/App.tsx [hunk 5]
```diff
@@ -1,14 +1,79 @@
...
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
+import { Todo } from './types/Todo';
+import { getTodos } from './api';
+
+function getFilteredTodos(key: string, query: string, todosToFilter: Todo[]) {
+  let filteredTodos = [];
+
+  switch (key) {
+    case 'all': {
```
Comment (Moroz-Dmytro) [inline src/App.tsx:17]: Create `enum` for these values

File: src/App.tsx [hunk 6]
```diff
@@ -17,18 +82,34 @@ export const App: React.FC = () => {
...
             <div className="block">
-              <Loader />
-              <TodoList />
+              {isLoading && <Loader />}
+              {!isLoading && (
+                <TodoList
+                  todos={filteredTodos}
+                  onClick={handleTodoSelection}
+                  selectedTodo={selectedTodo}
+                />
```
Comment (Moroz-Dmytro) [inline src/App.tsx:100]: Use ternary here

File: src/components/TodoList/TodoList.tsx [hunk 7]
```diff
@@ -1,100 +1,84 @@
...
+              {todo.completed
+                ? (
+                  <td className="is-vcentered">
+                    <span className="icon" data-cy="iconCompleted">
+                      <i className="fas fa-check" />
+                    </span>
+                  </td>
+                ) : (
+                  <td className="is-vcentered" />
+                )}
```
Comment (Moroz-Dmytro) [inline src/components/TodoList/TodoList.tsx:51]: DRY. `<td className="is-vcentered" />` is the same for both variants

**Review by Moroz-Dmytro — CHANGES_REQUESTED** [review]
I guess, you forgot to do deploy again, because I don't see changes after your fixes

File: src/App.tsx (hunk 5 above)
Comment (roman-mirzoian) [reply src/App.tsx:20]: Not fixed 😢

File: src/App.tsx [hunk 8]
```diff
@@ -1,14 +1,85 @@
...
+    setQuery('');
+  }, []);
+
+  useEffect(() => {
+    setIsLoading(true);
+    getTodos()
+      .then(todosFromServer => {
+        setTodos(todosFromServer);
+      })
+      .catch(() => setTodos([]))
```
Comment (roman-mirzoian) [inline src/App.tsx:79]: In the `catch()` block, it is necessary to catch an error in order to understand what went wrong.
Fix in all code.

File: src/components/TodoList/TodoList.tsx (hunk 7 above)
Comment (roman-mirzoian) [reply src/components/TodoList/TodoList.tsx:51]: Not fixed 😢

File: src/components/TodoModal/TodoModal.tsx [hunk 9]
```diff
@@ -15,35 +36,44 @@ export const TodoModal: React.FC = () => {
...
-              Todo #2
+              {`Todo #${selectedTodo.id}`}
             </div>
 
             {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
             <button
               type="button"
               className="delete"
               data-cy="modal-close"
+              onClick={() => onClick(null)}
```
Comment (roman-mirzoian) [inline src/components/TodoModal/TodoModal.tsx:47]: From the name `onClick(null)` it is not very clear what will happen after the click.

File: src/components/TodoFilter/TodoFilter.tsx [hunk 10]
```diff
@@ -16,19 +33,24 @@ export const TodoFilter = () => (
...
-        />
-      </span>
+      {!!query && (
+        <span className="icon is-right" style={{ pointerEvents: 'all' }}>
+          {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
+          <button
+            data-cy="clearSearchButton"
+            type="button"
+            className="delete"
+            onClick={onDelete}
```
Comment (roman-mirzoian) [inline src/components/TodoFilter/TodoFilter.tsx:50]: From the name onDelete it is not very clear what will delete after the click.

**Review by roman-mirzoian — CHANGES_REQUESTED** [review]
Do not ignore comments from the mentor: if you do not agree with the comment - write it, if you did not understand the comment - ask the mentor in the personal channel what he meant.
Also, now when opening information about the place, the modal window flashes, try to fix it, if you can't - ask for help:
https://gyazo.com/c1bd1c7720463aab7cf9392f54de0636

**Review by alexander-ignatow — APPROVED** [review]
LGTM!

**Pull request closed without merging**